### ISSUE TITLE
'INT_MAX' undeclared

### ISSUE THREAD

**Issue by jungle-boogie** [Tarsnap/tarsnap#249]
Hi All,

Attempting to build `tarsnap` from the git repo on FreeBSD 11-release.

```
% gmake
gmake  all-am
gmake[1]: Entering directory '/usr/home/jungle/bin/tarsnap'
gcc -std=gnu99 -DHAVE_CONFIG_H -I.  -I./tar -I./libarchive -I./lib/crypto -I./lib/datastruct -I./lib/keyfile -I./lib/netpacket -I./lib/netproto -I./lib/network -I./lib/scryptenc -I./lib/util -I./libcperciva/alg -I./libcperciva/cpusupport -I./libcperciva/crypto -I./libcperciva/datastruct -I./libcperciva/events -I./libcperciva/network -I./libcperciva/util -DCPUSUPPORT_CONFIG_FILE=\"cpusupport-config.h\" -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=700 -DPOSIXFAIL_MSG_NOSIGNAL -DPOSIXFAIL_CLOCK_REALTIME   -g -O2 -Wall -MT lib/crypto/lib_libtarsnap_a-crypto_keys_subr.o -MD -MP -MF lib/crypto/.deps/lib_libtarsnap_a-crypto_keys_subr.Tpo -c -o lib/crypto/lib_libtarsnap_a-crypto_keys_subr.o `test -f 'lib/crypto/crypto_keys_subr.c' || echo './'`lib/crypto/crypto_keys_subr.c
lib/crypto/crypto_keys_subr.c: In function 'import_BN':
lib/crypto/crypto_keys_subr.c:60:12: error: 'INT_MAX' undeclared (first use in this function)
  if (len > INT_MAX) {
            ^
lib/crypto/crypto_keys_subr.c:60:12: note: each undeclared identifier is reported only once for each function it appears in
gmake[1]: *** [Makefile:2288: lib/crypto/lib_libtarsnap_a-crypto_keys_subr.o] Error 1
gmake[1]: Leaving directory '/usr/home/jungle/bin/tarsnap'
gmake: *** [Makefile:1153: all] Error 2
```

I see this commit from a little more than a month ago:
https://github.com/Tarsnap/tarsnap/commit/95132525ab533cd6d52ddcd83ccad2d397a2c751

thanks!

**Comment by gperciva** [comment]
Interesting!  I see that `INT_MAX` is defined in `<limits.h>` in POSIX, so I've made a PR to fix this.  However, I can't reproduce with my own FreeBSD installation.  Could you please post the result of
```
uname -a
```

on your system?

**Comment by jungle-boogie** [comment]
yes, sir!

```
sean@lyliechild[15:57] ~/bin/baresip % uname -a
FreeBSD gperciva .in.tarsnap.net 11.0-RELEASE-p1 FreeBSD 11.0-RELEASE-p1 #0 r306420: Thu Sep 29 03:40:55 UTC 2016     root@releng2.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  i386
sean@lyliechild[15:58] ~/bin/baresip % freebsd-version -ku
11.0-RELEASE-p9
11.0-RELEASE-p9
```

**Comment by gperciva** [comment]
Thanks!  Unfortunately that hasn't clarified why I don't see that compile error on my systems, but I don't know much about FreeBSD so I can easily believe that I'm missing something.  Maybe @cperciva can shed some light?

**Comment by cperciva** [comment]
Looks like the OpenSSL in 11.0-RELEASE has <limits.h> included from <openssl/bn.h>.

Is @jungle-boogie using a different SSL library?

**Comment by jungle-boogie** [comment]
hi @cperciva,

yes, I'm using LibreSSL 2.5.3.

Thanks,
jb

**Comment by gperciva** [comment]
I believe this is now fixed.

**Comment by jungle-boogie** [comment]
Hi @gperciva,

I performed the following:
```
git pull (now on commit 11d07ea576fcec75e987abf01c48bda1e7282454)
make clean
autoreconf -i
./configure
```

The outcome:
```
UPPORT_CONFIG_FILE=\"cpusupport-config.h\"  -D_POSIX_C_SOURCE=200809L            -D_XOPEN_SOURCE=700                     -DPOSIXFAIL_MSG_NOSIGNAL                -DPOSIXFAIL_CLOCK_REALTIME   -g -O2 -Wall -MT lib/crypto/lib_libtarsnap_a-crypto_rsa.o -MD -MP -MF lib/crypto/.deps/lib_libtarsnap_a-crypto_rsa.Tpo -c -o lib/crypto/lib_libtarsnap_a-crypto_rsa.o `test -f 'lib/crypto/crypto_rsa.c' || echo './'`lib/crypto/crypto_rsa.c
In file included from lib/crypto/crypto_rsa.c:3:0:
lib/crypto/crypto_rsa.c: In function 'crypto_rsa_verify':
lib/crypto/crypto_rsa.c:173:18: error: 'INT_MAX' undeclared (first use in this function)
  assert(siglen < INT_MAX);
                  ^
lib/crypto/crypto_rsa.c:173:18: note: each undeclared identifier is reported only once for each function it appears in
lib/crypto/crypto_rsa.c: In function 'crypto_rsa_decrypt':
lib/crypto/crypto_rsa.c:387:15: error: 'INT_MAX' undeclared (first use in this function)
  assert(len < INT_MAX);
               ^
*** Error code 1

Stop.
make[1]: stopped in /usr/home/jungle/bin/tarsnap
*** Error code 1

Stop.
make: stopped in /usr/home/jungle/bin/tarsnap
26.572 real, 10.958 user, 1.582 sys;  page: 38 hard/139077 soft, swap: 0, I/O: 10/56
Mem: 27248KB (26950KB shared + 1651KB data/stack = 28601KB), VCSW: 1258 IVCSW: 2146
```

**Comment by gperciva** [comment]
Thanks for the report!  I have a fix for this specific issue, but there's other problem that arise with libressl.  I'm investigating now so that I can discuss them with @cperciva.

**Comment by jungle-boogie** [comment]
Hi @cperciva,

Alas

**Comment by gperciva** [comment]
Hi @jungle-boogie, I believe that this was fixed in https://github.com/Tarsnap/tarsnap/pull/252.

**Comment by jungle-boogie** [comment]
Hi @gperciva,

I don't know if this is related to that fix, or something new but I have this as a failure on freeBSD 11-release:
```
/multitape/tarsnap-multitape_chunkiter.o  tar/multitape/tarsnap-multitape_transaction.o  tar/chunks/tarsnap-chunks_write.o  tar/chunks/tarsnap-chunks_read.o  tar/chunks/tarsnap-chunks_stats.o  tar/chunks/tarsnap-chunks_delete.o  tar/chunks/tarsnap-chunks_transaction.o  tar/chunks/tarsnap-chunks_stats_internal.o  tar/chunks/tarsnap-chunks_directory.o  tar/storage/tarsnap-storage_write.o  tar/storage/tarsnap-storage_read.o  tar/storage/tarsnap-storage_delete.o  tar/storage/tarsnap-storage_directory.o  tar/storage/tarsnap-storage_transaction.o    libarchive/libarchive.a lib/libtarsnap.a lib/libtarsnap_aesni.a  lib/libtarsnap_sse2.a -llzma -lbz2 -lcrypto -lz -lrt
tar/multitape/tarsnap-multitape_metadata.o: In function `multitape_metadata_enc':
/usr/home/sean/bin/tarsnap/tar/multitape/multitape_metadata.c:72: undefined reference to `assert'
collect2: error: ld returned 1 exit status
*** Error code 1

Stop.
make[1]: stopped in /usr/home/sean/bin/tarsnap
*** Error code 1

Stop.
make: stopped in /usr/home/sean/bin/tarsnap
```

This is the failure on OpenBSD -current:
```
libarchive/archive_entry.c:1526: warning: warning: wcscpy() is almost always misused, please use wcslcpy()                                                             tar/tarsnap-bsdtar.o: In function `main':                                                                                                                              tar/bsdtar.c:908: warning: warning: strcpy() is almost always misused, please use strlcpy()                                                                            tar/multitape/tarsnap-multitape_metadata.o: In function `multitape_metadata_enc':                                                                                      tar/multitape/multitape_metadata.c:72: undefined reference to `assert'                                                                                                 collect2: ld returned 1 exit status                                                                                                                                    *** Error 1 in . (Makefile:1883 'tarsnap')
*** Error 1 in /home/sean/bin/tarsnap (Makefile:1153 'all')
```

**Comment by gperciva** [comment]
Hi @jungle-boogie: yikes, this was unfortunate timing!  My apologies.  The `assert()` problem arose from half a dozen unrelated PRs being merged late last night (fix in #265), and it appears that Linux is a bit more lax about includes than FreeBSD.

I suspect that OpenBSD has the same problem; the `wcspy()` message looks like a warning rather than an error.

Sorry about this.  Once the current round of PRs are in, I'll do another set of testing and let you know if I think it's ok.

**Comment by jungle-boogie** [comment]
Hi @gperciva,

That's not a problem at all. Thanks for the update!

**Comment by gperciva** [comment]
Hi @jungle-boogie,
I believe this is fixed in current git master 7f178c74513b5adfccda4be0f52c600da34733bc.  I've tested it on OpenBSD 64-bit (with libressl), and freebsd-i386 (no libressl).  Thanks for your patience!

**Comment by jungle-boogie** [comment]
Hi @gperciva,

Yes, all better! Thanks for getting this resolved and tracking down the issues.
